### ISSUE TITLE
Wazuh-db error found in Manager in Footprint tests

### ISSUE THREAD

**Issue by hossam1522** [wazuh/wazuh#25918]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.9.1-RC 1 | wazuh-db | Manager | Packages/Sources | OS version |

While analyzing logs in https://github.com/wazuh/wazuh/issues/25880, it was found in the manager logs the following error related to wazuh-db:

```
2024/09/22 23:01:34 wazuh-db: ERROR: SQLite: UNIQUE constraint failed: sca_scan_info.id
```

Further research is required.

**Comment by juliancnn** [comment]
## Update: Root Cause Analysis

```plaintext
2024/09/22 23:01:34 wazuh-db: ERROR: SQLite: UNIQUE constraint failed: sca_scan_info.id
```

The error arises because an attempt is being made to insert a value into the `sca_scan_info` table that already exists in the `id` column, which is a primary key, preventing the insertion of duplicate values. The error likely originates from the following code snippet, which is the only `insert` operation into `sca_scan_info` that I have found.

https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb_sca.c#L505-L537

Where [WDB_STMT_SCA_SCAN_INFO_INSERT](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb.c#L110-L111) is:
```c
    [WDB_STMT_SCA_SCAN_INFO_INSERT] = "INSERT INTO sca_scan_info (start_scan,end_scan,id,policy_id,pass,fail,invalid,total_checks,score,hash) VALUES (?,?,?,?,?,?,?,?,?,?);",
```

Analysis of the code sequence followed by wazuh-db that leads to this error:

- **run_worker (src/wazuh_db/main.c:385)**: receives a message via the wazuh-db socket.
- **run_worker (src/wazuh_db/main.c:419)**: Calls `wdb_parse` with the received message, including the query.
- **wdb_parse (src/wazuh_db/wdb_parser.c:338)**: Detects it is an agent query, as the received query must be `agent id_agent sca insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse (src/wazuh_db/wdb_parser.c:374)**: Detects it is an SCA type query.
- **wdb_parse (src/wazuh_db/wdb_parser.c:383)**: Performs the query `insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2011)**: Receives the query `insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2622)**: Parses the `insert_scan_info` query type, which already includes the ID.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2676)**: Retrieves the already existing ID.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2767)**: Calls `wdb_sca_scan_info_save` with the existing ID.
- **wdb_sca_scan_info_save (src/wazuh_db/wdb_sca.c:505)**: Receives a parameter with an ID that already exists in the `sca_scan_info` table.
- **wdb_sca_scan_info_save/ calls wdb_step (src/wazuh_db/wdb_sca.c:532)**: Executes the query and it fails.

We can see that the query is already received from an external program through the socket. The responsibility for sending the query with a duplicated ID lies with wazuh-analysisd in the SCA decoders, here:

https://github.com/wazuh/wazuh/blob/38483be2d5834ddf2d6e0e1d86c015dcc11334ce/src/analysisd/decoders/security_configuration_assessment.c#L626-L655

The function `SaveScanInfo` already receives a duplicated ID and simultaneously `update` is marked as false, which leads to an attempt to insert a duplicate value into the `sca_scan_info` table instead of performing an `update_scan_info_start`.

To determine whether to update, a query to wazuh-db is made here:

https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L1054-L1055

At this point, it checks if there is a scan with the same ID; if there is, an update should be made; if not, an insert should be performed.

This situation could potentially create a race condition under very rare circumstances between processes and threads. Multiple `w_decode_sca_thread` in wazuh-analysisd process SCA messages and compete to query wazuh-db to check whether the scan_id already exists or not and to insert the row into the `sca_scan_info` table. Meanwhile, wazuh-db may have multiple threads processing queries from wazuh-analysisd, with no guarantee that responses to each query are in order.

**Comment by juliancnn** [comment]
## Update: Root Cause Analysis

```plaintext
2024/09/22 23:01:34 wazuh-db: ERROR: SQLite: UNIQUE constraint failed: sca_scan_info.id
```

The error arises because an attempt is being made to insert a value into the `sca_scan_info` table that already exists in the `id` column, which is a primary key, preventing the insertion of duplicate values. The error likely originates from the following code snippet, which is the only `insert` operation into `sca_scan_info` that I have found.

https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb_sca.c#L505-L537

Where [WDB_STMT_SCA_SCAN_INFO_INSERT](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb.c#L110-L111) is:
```c
    [WDB_STMT_SCA_SCAN_INFO_INSERT] = "INSERT INTO sca_scan_info (start_scan,end_scan,id,policy_id,pass,fail,invalid,total_checks,score,hash) VALUES (?,?,?,?,?,?,?,?,?,?);",
```

Analysis of the code sequence followed by wazuh-db that leads to this error:

- **run_worker (src/wazuh_db/main.c:385)**: receives a message via the wazuh-db socket.
- **run_worker (src/wazuh_db/main.c:419)**: Calls `wdb_parse` with the received message, including the query.
- **wdb_parse (src/wazuh_db/wdb_parser.c:338)**: Detects it is an agent query, as the received query must be `agent id_agent sca insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse (src/wazuh_db/wdb_parser.c:374)**: Detects it is an SCA type query.
- **wdb_parse (src/wazuh_db/wdb_parser.c:383)**: Performs the query `insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2011)**: Receives the query `insert_scan_info | pm_start_scan | pm_end_scan | scan_id(DUPLICATED ID) | .....`.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2622)**: Parses the `insert_scan_info` query type, which already includes the ID.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2676)**: Retrieves the already existing ID.
- **wdb_parse_sca (src/wazuh_db/wdb_parser.c:2767)**: Calls `wdb_sca_scan_info_save` with the existing ID.
- **wdb_sca_scan_info_save (src/wazuh_db/wdb_sca.c:505)**: Receives a parameter with an ID that already exists in the `sca_scan_info` table.
- **wdb_sca_scan_info_save/ calls wdb_step (src/wazuh_db/wdb_sca.c:532)**: Executes the query and it fails.

We can see that the query is already received from an external program through the socket. The responsibility for sending the query with a duplicated ID lies with wazuh-analysisd in the SCA decoders, here:

https://github.com/wazuh/wazuh/blob/38483be2d5834ddf2d6e0e1d86c015dcc11334ce/src/analysisd/decoders/security_configuration_assessment.c#L626-L655

The function `SaveScanInfo` already receives a duplicated ID and simultaneously `update` is marked as false, which leads to an attempt to insert a duplicate value into the `sca_scan_info` table instead of performing an `update_scan_info_start`.

To determine whether to update, a query to wazuh-db is made here:

https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L1054-L1055

At this point, it checks if there is a scan with the same ID; if there is, an update should be made; if not, an insert should be performed.

This situation could potentially create a race condition under very rare circumstances between processes and threads. Multiple `w_decode_sca_thread` in wazuh-analysisd process SCA messages and compete to query wazuh-db to check whether the scan_id already exists or not and to insert the row into the `sca_scan_info` table. Meanwhile, wazuh-db may have multiple threads processing queries from wazuh-analysisd, with no guarantee that responses to each query are in order.

----------------------
## Update 2: Another RCA

Another possible cause is that there really was a need to attempt an insert, because the decision to either update or insert is made solely by looking at the `policy_id`:

Here:

[https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L1054-L1091](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L1054-L1091)

and Here:

[https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L313-L314](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/analysisd/decoders/security_configuration_assessment.c#L313-L314)

Then in wazuh-db:

[https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb_sca.c#L304-L334](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/wdb_sca.c#L304-L334)

Where:

```c
    [WDB_STMT_SCA_FIND_SCAN] = "SELECT policy_id,hash,id FROM sca_scan_info WHERE policy_id = ?;",
```

The problem lies in that the table has `id` as the primary key:

[https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/schema_upgrade_v2.sql#L20-L32](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_db/schema_upgrade_v2.sql#L20-L32)

And this field is generated as a random number on the agent, here:

[https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_modules/wm_sca.c#L432-L445](https://github.com/wazuh/wazuh/blob/8e0e581a1925a90472fd6a7bdbc358947f809e3f/src/wazuh_modules/wm_sca.c#L432-L445)

Considering this number is 32-bits and if negative it is converted to positive, then the range is from 0 to 2^31-1.

On the other hand, in this test, the manager conducts scans every second (although the default configuration is every 12 hours), implying that during the 36-hour duration of the test, 129,600 scans were generated.

We can then apply the birthday paradox approximation to calculate the probability that there is a duplicate among 129,600 random 31-bit numbers:

```python
import math

def birthday_paradox(n, m):
    return 1 - math.exp(-m*(m-1)/(2*n))

n = 2**31
m = 129600

birthday_paradox(n, m)
```

The probability of having a duplicate in 129,600 random 31-bit numbers is 0.9799721318708781, making it highly probable that a duplicate ID was generated by the agents.

If it happens that a duplicate was generated, then an attempt to insert into the `sca_scan_info` table fails due to the `id` field being a primary key, resulting in the error `UNIQUE constraint failed: sca_scan_info.id`.

**Comment by Dwordcito** [comment]
The problem is pre-existing, therefore it is not a blocker for 4.9.1.

Apart from this, and given Julian's explanation, this is a consequence of the poor choice of the pk and the SCA configuration, which by default is 12 hours and in this tests it is one second, when this happens, the randomization can cause this.

The effect is null when you have only one second of configuration, because in the next scan with a different ID (after the second) it will be satisfactory.


This will be fixed in 5.0 as SCA dbsync will be integrated to avoid these problems.

All this leads me to the fact that we will not work on it and I will close it as not planned due to low prevalence and null effect.
